### PR TITLE
[Inductor] auto-chunker

### PR DESCRIPTION
**Pull request by shunting314**
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* #171359
* __->__ #136702
* #170649

The AutoChunker defines the following chunking metadata and propagate them thru the subgraphs that we chunk
1. scale_by: The AutoChunker is only enabled if there is a single scalar tangent. To decouple the dependency on tangent for the bwd subgraph so we can compute them in fwd, we pretend the tangent is 1 first and record it in 'scale_by' . This metadata get propagated and when we cancel the chunking effect in the end of bwd subgraph, we apply the scaling.
2. chunk_dim: record which dimension of the tensor get chunked
3. need_sum: if true, the original Tensor is the sum (rather than concat) of each chunked tensors.

One important implementation detail is, we need put chunked subgraph in a HOP (use invoke_subgraph here). Otherwise Inductor fuse across these subgraphs and results in no peak memory saving.

Here are some early benchmarking result on GPT2.
- 64 chunks:
   - final 19 iters avg: 242.550ms
   - peak memory consumption: 12603 MiB
- 32 chunks:
   - final 19 iters avg: 206.180ms
   - peak memory consumption: 12880 MiB
- 16 chunks
   - final 19 iters avg: 196.997ms
   - peak memory consumption: 13267 MiB
- 8 chunks
   - final 19 iters avg: 194.924ms
   - peak memory consumption: 14049 MiB

With 64 chunks, our peak memory is smaller than llm.c's 13.4GB.

I also tried the AutoChunker on PT2 OSS benchmarks to verify the numerical. By default our accuracy test picks a very small batch size. This makes AutoChunker get skipped. I force batch_size to be 16 for BertForMaskedLM to trigger the AutoChunker and verified the numerical correctness.

cc @voznesenskym @penguinwu @EikanWang @jgong5 @Guobing-Chen @XiaobingSuper @zhuhaozhe @blzheng @wenzhe-nrv @jiayisunx @ipiszy @kadeng @muchulee8 @amjames @chauhang @aakhundov @coconutruben @jataylo @Lucaskabela @yf225 @chenyang78 @ColinPeppler @desertfire @ngimel

### CODE REVIEW

File: torch/_inductor/fx_passes/auto_chunker/__init__.py [hunk 1]
```diff
@@ -0,0 +1,62 @@
...
+        raise CantChunk("Skip chunking due to no amplifier node found")
+
+    propagate(amplifier_node)
+    if not tangent_has_chunking_meta(gm):
+        raise CantChunk(
+            "Skip chunking either because the graph is for inference only or "
+            "because the chunking metadata does not propagate to the backward "
+            "(e.g. due to too trivial loss function)"
+        )
+    out_gm = ChunkingApplier(gm, config.AutoChunker.num_chunk or 8).apply()
```
Comment (shunting314) [inline torch/_inductor/fx_passes/auto_chunker/__init__.py:60]: Right now this PR defaults to use 8 chunks.

I'll do in a follow up PR to make it better by
- either use some heuristics to create as many chunks as possible until the M dimension of matmul is smaller than some threshold
- or build some autotuning functionality

**Review by jansel — CHANGES_REQUESTED** [review]
Looks like tests are failing, re-request review once you want me to look in more detail.

File: test/inductor/test_auto_chunker.py [hunk 2]
```diff
@@ -0,0 +1,170 @@
...
+        self.assertTrue(
+            peak_memory < expected_bound,
+            f"Actual peak_memory {peak_memory}, expected bound {expected_bound}",
+        )
+
+
+if __name__ == "__main__":
+    from torch._inductor.test_case import run_tests
+
+    if HAS_GPU:
```
Comment (etaf) [inline test/inductor/test_auto_chunker.py]: Hi, since the test cases are all hard coded `cuda`, suggest to not use `HAS_GPU` which allows other GPU run this test case and then failed with no `cuda` support. 
```suggestion
    if HAS_CUDA:
```

**Review by etaf — DISMISSED** [review]

File: test/inductor/test_auto_chunker.py [hunk 3]
```diff
@@ -0,0 +1,170 @@
+# Owner(s): ["module: inductor"]
+import os
+
+import torch
+from torch import nn
+from torch._dynamo.utils import same
+from torch._inductor import config, metrics
+from torch._inductor.test_case import TestCase
+from torch.testing._internal.common_device_type import largeTensorTest
+from torch.testing._internal.inductor_utils import GPU_TYPE, HAS_GPU
```
Comment (etaf) [inline test/inductor/test_auto_chunker.py]:
```suggestion
from torch.testing._internal.inductor_utils import GPU_TYPE, HAS_CUDA
```

File: torch/_inductor/config.py [hunk 4]
```diff
@@ -946,6 +946,22 @@ def decide_compile_threads() -> int:
 enable_caching_generated_triton_templates: bool = False
 
 
+class AutoChunker:
```
Comment (jansel) [inline torch/_inductor/config.py]: match naming convention elsewhere in config file

File: torch/_inductor/fx_passes/auto_chunker/__init__.py [hunk 5]
```diff
@@ -0,0 +1,65 @@
...
+    if gm.meta.get("produced_by_chunker", False):
+        # Don't chunk a graph produced by the chunker
+        return gm
+
+    if len(get_tangent_nodes(gm.graph)) == 0:
+        # no tangents. Can be the optimizer graph
+        # skip
+        return gm
+
+    log.debug("Joint graph before chunking:\n%s", gm.print_readable(False))
```
Comment (jansel) [inline torch/_inductor/fx_passes/auto_chunker/__init__.py]: Avoid calling print_readable if debug mode not enabled.

File: torch/_inductor/fx_passes/auto_chunker/applier.py [hunk 6]
```diff
@@ -0,0 +1,423 @@
+# mypy: allow-untyped-defs
+# mypy: disable-error-code=union-attr
```
Comment (jansel) [inline torch/_inductor/fx_passes/auto_chunker/applier.py]: Type this file

File: torch/_inductor/fx_passes/auto_chunker/core.py [hunk 7]
```diff
@@ -0,0 +1,216 @@
+# mypy: allow-untyped-defs
```
Comment (jansel) [inline torch/_inductor/fx_passes/auto_chunker/core.py]: Type this file

File: torch/_inductor/fx_passes/auto_chunker/core.py [hunk 8]
```diff
@@ -0,0 +1,216 @@
...
+    r"""
+    Find the 'amplifier' node which is a not that generates large
+    output with small/medium input.
+
+    If there are multiple amplifier nodes, return the one with the largest
+    amplification ratio.
+    """
+
+    amplifier_nodes_ratio = []
+    for node in graph.nodes:
```
Comment (jansel) [inline torch/_inductor/fx_passes/auto_chunker/core.py]: Do we need to iterate over the graph some many times?  Any place we could use `graph.find_nodes()` instead?

File: torch/_inductor/fx_passes/auto_chunker/propagate_scale_by.py [hunk 9]
```diff
@@ -0,0 +1,166 @@
+# mypy: allow-untyped-defs
```
Comment (jansel) [inline torch/_inductor/fx_passes/auto_chunker/propagate_scale_by.py]: Type this file

File: torch/_inductor/fx_passes/auto_chunker/propagator.py [hunk 10]
```diff
@@ -0,0 +1,629 @@
+# mypy: allow-untyped-defs
```
Comment (jansel) [inline torch/_inductor/fx_passes/auto_chunker/propagator.py]: Same

**Review by jansel — CHANGES_REQUESTED** [review]
Have we done a full CI run and benchmark run with this enabled?

**Comment by shunting314** [comment]
> Have we done a full CI run and benchmark run with this enabled?

Not yet. I'll kick off one. I may need tweak the batch size for accuracy test a bit. Right now it's very small and auto-chunker will be skipped in that case. I'll try to override the batch size so we can test for accuracy with auto-chunker taking effect

**Comment by shunting314** [comment]
@v0i0 thanks for offering review. That would be great!

I just need find bandwidth to resolve comments, do more tests and get this in.

File: torch/_inductor/fx_passes/auto_chunker/__init__.py (hunk 5 above)
Comment (v0i0) [reply torch/_inductor/fx_passes/auto_chunker/__init__.py]: maybe just wrap with GraphTransformObserver instead?

File: torch/_inductor/fx_passes/auto_chunker/propagator.py [hunk 11]
```diff
@@ -0,0 +1,629 @@
...
+        aten.add.Tensor,
+        aten.sub.Tensor,
+        aten.div.Tensor,
+        aten.mul.Tensor,
+        prims.fma.default,
+        aten.where.self,
+        aten.neg.default,
+    ]
+)
+def propagate_general_copy_metadata(out_node, ignore_broadcast=False):
```
Comment (v0i0) [inline torch/_inductor/fx_passes/auto_chunker/propagator.py]: while i admittedly have no idea how to do it, this seems like a special case of propagating along vmap/batching rules. if they exist somewhere, you could probably cut this down a lot.

**Review by v0i0 — COMMENTED** [review]
this is amazing! what does the generated code look like? why is this a bunch of invoke_subgraph rather than vmap w/ chunk_size or hop.scan?

File: torch/_inductor/fx_passes/auto_chunker/core.py (hunk 8 above)
Comment (shunting314) [reply torch/_inductor/fx_passes/auto_chunker/core.py]: This loop is absolutely unneeded. Removed.

File: test/inductor/test_flex_attention.py [hunk 12]
```diff
@@ -4438,8 +4438,8 @@ def forward(self, arg0_1: "f64[]", arg1_1: "i32[]", arg2_1: "i32[]", arg3_1: "i3
 
     class mask_graph0(torch.nn.Module):
         def forward(self, arg0_1: "i32[]", arg1_1: "i32[]", arg2_1: "i32[]", arg3_1: "i32[]"):
-            full_default: "b8[]" = torch.ops.aten.full.default([], True, dtype = torch.bool, layout = torch.strided, device = device(type='GPU_TYPE', index=0), pin_memory = False)
-            return full_default
+            full_default_1: "b8[]" = torch.ops.aten.full.default([], True, dtype = torch.bool, layout = torch.strided, device = device(type='GPU_TYPE', index=0), pin_memory = False)
+            return full_default_1
```
Comment (jansel) [inline test/inductor/test_flex_attention.py]: What is cuasing this change?

File: torch/_inductor/compile_fx.py [hunk 13]
```diff
@@ -527,25 +527,40 @@ def _recursive_pre_grad_passes(
 
 def _recursive_joint_graph_passes(
     gm: GraphModule, skip_invoke_subgraph: bool = False
-) -> None:
+) -> GraphModule:
```
Comment (jansel) [inline torch/_inductor/compile_fx.py:530]: Why change the return value?  It looks like we are still mutating `gm` not creating a new GraphModule?

File: torch/_inductor/fx_passes/auto_chunker/__init__.py [hunk 14]
```diff
@@ -0,0 +1,81 @@
...
+    The chunking operation is propagated thru the fx graph until a point we should
+    re-generate non-chunked tensors.
+
+    Only chunk across the batch dimension of the tensor for now.
+    """
+    graph = gm.graph
+
+    if torch._inductor.config.cpp_wrapper:
+        # cpp wrapper does not support codegening invoke_subgraph
+        return gm
```
Comment (jansel) [inline torch/_inductor/fx_passes/auto_chunker/__init__.py]: Should this be an error?

File: torch/_inductor/fx_passes/auto_chunker/propagate_scale_by.py [hunk 15]
```diff
@@ -0,0 +1,196 @@
...
+    get_args_of_node_type,
+    get_scale_by_from_metas,
+    get_scale_by_from_node,
+)
+
+
+aten = torch.ops.aten
+prims = torch.ops.prims
+
+propagate_rules = {}  # type: ignore[var-annotated]
```
Comment (jansel) [inline torch/_inductor/fx_passes/auto_chunker/propagate_scale_by.py]: Annotate types?

File: torch/_inductor/fx_passes/auto_chunker/propagator.py [hunk 16]
```diff
@@ -0,0 +1,700 @@
+# pyrefly: ignore-errors
+
+"""
+We ignore pyrefly errors on this files since a lot of errors reported by
+pyre does not make sense. E.g. it complains the arg type for
+    if not has_any_chunking_meta(bias_node, input_node, weight_node):
+while we already assert that all the inputs are of Node type.
+mypy does not have this issue.
+"""
```
Comment (jansel) [inline torch/_inductor/fx_passes/auto_chunker/propagator.py]: IMO we should get the types working properly.

**Review by jansel — COMMENTED** [review]
Is the `doctr_det_predictor` failure real?

File: torch/_inductor/compile_fx.py (hunk 13 above)
Comment (shunting314) [reply torch/_inductor/compile_fx.py:530]: The ChunkingApplier would create a new GraphModule: 
```
        newgm = torch.fx._lazy_graph_module._make_graph_module(
            self.gm,
            self.parent_graph,
        )
```

I initially decided to create a new GraphModule since it's easier to fallback.

File: torch/_inductor/fx_passes/auto_chunker/__init__.py (hunk 14 above)
Comment (shunting314) [reply torch/_inductor/fx_passes/auto_chunker/__init__.py]: yea. I'll add a error message and fallback to not chunking

File: test/inductor/test_flex_attention.py (hunk 12 above)
Comment (shunting314) [reply test/inductor/test_flex_attention.py]: It turns out to be a bug:

```
        old_subgraph_names = _get_subgraph_names(gm, skip_invoke_subgraph)
        for subgraph_name in old_subgraph_names:
            _run_on_sub_graph_module(subgraph_name)

        out_gm = joint_graph_passes(gm)

        # Some joint graph passes may create new sub graph module. Run one round
        # for the newly created graph modules.
        # We should not skip graphs for invoke_subgraph HOPs for newly
        # generated subgraphs.
        for subgraph_name in _get_subgraph_names(out_gm, skip_invoke_subgraph=False):
            if subgraph_name not in old_subgraph_names:
                _run_on_sub_graph_module(subgraph_name)
```

The first line returns a Python generator. It's iterated once in the first for loop. When it's used in the second for-loop, it's already empty. This cause joint-graph passes being applied to sub-graph a second time and triggers the codegen change.

Fixed it in the next commit.

File: torch/_inductor/fx_passes/auto_chunker/propagator.py (hunk 16 above)
Comment (shunting314) [reply torch/_inductor/fx_passes/auto_chunker/propagator.py]: Looks like pyre does not 'see' lines like 'assert isinstance(bias_node, Node)' in the enclosing function (mypy can do that).

I'll workaround this by moving the assertion to nested function.

**Review by jansel — APPROVED** [review]

**Comment by shunting314** [comment]
@pytorchbot merge -i

A comment about the PR size. The PR has more than 2K lines and violate a lint. TBH, all the code here are necessary to make  a workable version of AutoChunker. If people have strong option for splitting it, I can move the test file to a separate PR, or land the PR module by module but test in the last PR.

**Comment by huydhn** [comment]
@pytorchbot revert -m 'Sorry for reverting your change but this seems to break vLLM CI' -c nosignal

[GH job link](https://github.com/pytorch/pytorch/actions/runs/20479091241/job/58851064679) [HUD commit link](https://hud.pytorch.org/pytorch/pytorch/commit/3cd6766108c76fac50de3deca135122ecc6837e1)
